### PR TITLE
[Function-NoOp] Implement Hash from Label

### PR DESCRIPTION
**Pull request by TheIronDev**
### Description

Given a hash label from a gcf function, apply it to the Backend's Endpoint.

### Scenarios Tested

- [x] Given a hash label, does it apply to Backend's Endpoint

### CODE REVIEW

File: src/gcp/cloudfunctions.ts [hunk 1]
```diff
@@ -559,6 +560,9 @@ export function endpointFromFunction(gcfFunction: CloudFunction): backend.Endpoi
     );
   }
   endpoint.codebase = gcfFunction.labels?.[CODEBASE_LABEL] || projectConfig.DEFAULT_CODEBASE;
+  if (gcfFunction.labels?.[HASH_LABEL]) {
+    endpoint.hash = gcfFunction.labels?.[HASH_LABEL];
```
Comment (colerogers) [inline src/gcp/cloudfunctions.ts:564]: Don't we also need the reverse logic to add the hash label to the function from the endpoint? Look inside `functionFromEndpoint()` on L527 for examples of how we do that with the other label types

Comment (taeold) [inline src/gcp/cloudfunctions.ts:564]: nit* since you already checked for existence of `HASH_LABEL`, I think you can omit the `?` (think TS is smart enough to deduce that).

**Review by taeold — DISMISSED** [review]

File: src/gcp/cloudfunctions.ts (hunk 1 above)
Comment (taeold) [reply src/gcp/cloudfunctions.ts:564]: +1. Is it being done in another PR?

File: src/deploy/functions/backend.ts [hunk 2]
```diff
@@ -359,6 +359,10 @@ export type Endpoint = TargetIds &
     // This is a temporary fix to address https://github.com/firebase/firebase-tools/issues/4171
     // GCFv1 can be http or https and GCFv2 is always https
     securityLevel?: gcf.SecurityLevel;
+
+    // "Hash" is a value derived from function labels that is used to check if there are any changes
+    // between the serverside and local copies of the function.
+    hash?: string;
```
Comment (inlined) [inline src/deploy/functions/backend.ts:365]: Sometimes I wish an Endpoint was a class so we could default fields like this to "" and not worry about nullish values.

File: src/gcp/cloudfunctions.ts (hunk 1 above)
Comment (inlined) [reply src/gcp/cloudfunctions.ts:564]: This should implement functionFromEndpoint and endpointFromFunction in both cloudfunctions.ts and cloudfunctionsv2.ts

Comment (inlined) [reply src/gcp/cloudfunctions.ts:564]: Yes, any time you're in the clause of an if statement that did a null check, the compiler stops requiring the ?. And in fact we should remove it to
A) Prove to ourselves that we've handled all edge cases
B) Ensure that this code, if copy + pasted, will only work when similarly guarded

File: src/test/gcp/cloudfunctions.spec.ts [hunk 3]
```diff
@@ -455,6 +455,30 @@ describe("cloudfunctions", () => {
...
     });
+
+    it("should derive hash from labels", () => {
+      expect(
+        cloudfunctions.endpointFromFunction({
+          ...HAVE_CLOUD_FUNCTION,
+          httpsTrigger: {},
+          labels: {
+            ...CLOUD_FUNCTION.labels,
+            [cloudfunctions.CODEBASE_LABEL]: "my-codebase",
```
Comment (inlined) [inline src/test/gcp/cloudfunctions.spec.ts:466]: In general I wonder if we should delete the labels in endpointFromFunction so that tests don't have to worry about both the label and the property implementation of these features when creating fake data.

File: src/gcp/cloudfunctions.ts (hunk 1 above)
Comment (TheIronDev) [reply src/gcp/cloudfunctions.ts]: Updated! Thanks for calling out!

Comment (TheIronDev) [reply src/gcp/cloudfunctions.ts]: Good catch! Updated!

File: src/test/gcp/cloudfunctions.spec.ts (hunk 3 above)
Comment (TheIronDev) [reply src/test/gcp/cloudfunctions.spec.ts]: I noticed this part. I was concerned about unintended side-effects, so I left the forced assignment as-is.

File: src/gcp/constants.ts [hunk 4]
```diff
@@ -0,0 +1,13 @@
...
+export const HASH_LABEL = "firebase-functions-hash";
+export const BLOCKING_LABEL = "deployment-blocking";
+export const BLOCKING_LABEL_KEY_TO_EVENT: Record<string, typeof AUTH_BLOCKING_EVENTS[number]> = {
+  "before-create": "providers/cloud.auth/eventTypes/user.beforeCreate",
+  "before-sign-in": "providers/cloud.auth/eventTypes/user.beforeSignIn",
+};
+export const BLOCKING_EVENT_TO_LABEL_KEY: Record<typeof AUTH_BLOCKING_EVENTS[number], string> = {
+  "providers/cloud.auth/eventTypes/user.beforeCreate": "before-create",
+  "providers/cloud.auth/eventTypes/user.beforeSignIn": "before-sign-in",
+};
```
Comment (colerogers) [inline src/gcp/constants.ts:13]: A downside for putting these here is that this directory is for all GCP services, not just functions. Maybe defining them in `src/functions` would be a better option. But @taeold might have some opinions about yanking these

**Review by colerogers — COMMENTED** [review]
LGTM, just the open issue about the constants file that @taeold should sign off on

File: src/gcp/constants.ts (hunk 4 above)
Comment (taeold) [reply src/gcp/constants.ts:13]: Agree with @colerogers here. Technically `gcp/cloudfunctions.ts` doesn't feel that great either, but at least that file is mainly used by us.

**Review by taeold — APPROVED** [review]
lgtm - mone minor comment on `gcp/constants.ts`.
